File: scripts/audit/audit.py
#!/usr/bin/env python3
"""
LOGOS Database Audit Script
Phase 0: Detect existing infrastructure and generate report
"""

import os
import sys
import json
import asyncio
from datetime import datetime
from typing import Dict, List, Any, Optional
from dataclasses import dataclass, asdict
import asyncpg

DATABASE_URL = os.getenv('DATABASE_URL',
    'postgresql://postgres:voqQePIH4adopQUa-1UUaFKnOT-mtsod@maglev.proxy.rlwy.net:49514/railway')

@dataclass
class AuditReport:
    timestamp: str
    embedding_infra: Dict[str, Any]
    source_counts: Dict[str, int]
    total_sources: int
    embedding_coverage: Dict[str, Any]
    translation_coverage: Dict[str, Any]
    morphology_coverage: Dict[str, int]
    backfill_jobs_exists: bool
    indexes: List[Dict[str, str]]
    existing_translation_tables: List[str]
    recommendations: List[str]

async def run_audit() -> AuditReport:
    """Run comprehensive database audit."""
    # Try connection with different SSL modes
    conn = None
    for ssl_mode in [False, 'prefer', 'require']:
        try:
            conn = await asyncpg.connect(DATABASE_URL, ssl=ssl_mode)
            print(f"Connected with ssl={ssl_mode}")
            break
        except Exception as e:
            print(f"Connection with ssl={ssl_mode} failed: {e}")
            continue

    if conn is None:
        raise Exception("Could not connect to database with any SSL mode")

    report = AuditReport(
        timestamp=datetime.now().isoformat(),
        embedding_infra={'columns': [], 'dimensions': {}, 'primary_store': 'unknown'},
        source_counts={},
        total_sources=0,
        embedding_coverage={'total': 0, 'with_embeddings': 0, 'missing': 0, 'coverage_percent': 0},
        translation_coverage={'total': 0, 'with_embeddings': 0, 'unique_urns': 0, 'multi_translation_urns': 0},
        morphology_coverage={'morph_entries': 0, 'passage_tokens': 0},
        backfill_jobs_exists=False,
        indexes=[],
        existing_translation_tables=[],
        recommendations=[]
    )

    try:
        print('Auditing embedding infrastructure...')

        # 1. Embedding columns
        embed_cols = await conn.fetch("""
            SELECT table_name, column_name, data_type
            FROM information_schema.columns
            WHERE table_schema = 'public'
            AND (data_type ILIKE '%vector%'
                 OR column_name ILIKE '%embed%'
                 OR column_name ILIKE '%vector%')
            ORDER BY table_name, column_name
        """)
        report.embedding_infra['columns'] = [
            {'table': r['table_name'], 'column': r['column_name'], 'type': r['data_type']}
            for r in embed_cols
        ]

        # 2. Check vector dimensions
        try:
            dims = await conn.fetchrow("""
                SELECT vector_dims(embedding) as dims
                FROM translations
                WHERE embedding IS NOT NULL
                LIMIT 1
            """)
            if dims:
                report.embedding_infra['dimensions']['translations'] = dims['dims']
        except Exception as e:
            print(f'Note: translations.embedding dimension check: {e}')

        try:
            dims = await conn.fetchrow("""
                SELECT
                  CASE WHEN original_embedding IS NOT NULL THEN vector_dims(original_embedding) ELSE NULL END as orig_dims,
                  CASE WHEN invariant_embedding IS NOT NULL THEN vector_dims(invariant_embedding) ELSE NULL END as inv_dims
                FROM style_invariant_embeddings
                WHERE original_embedding IS NOT NULL OR invariant_embedding IS NOT NULL
                LIMIT 1
            """)
            if dims:
                report.embedding_infra['dimensions']['style_invariant_original'] = dims['orig_dims']
                report.embedding_infra['dimensions']['style_invariant_invariant'] = dims['inv_dims']
        except Exception as e:
            print(f'Note: style_invariant_embeddings dimension check: {e}')

        try:
            dims = await conn.fetchrow("""
                SELECT vector_dims(embedding) as dims
                FROM passages
                WHERE embedding IS NOT NULL
                LIMIT 1
            """)
            if dims:
                report.embedding_infra['dimensions']['passages'] = dims['dims']
        except Exception as e:
            print(f'Note: passages.embedding dimension check: {e}')

        # 3. Source text counts by language
        print('Counting source texts...')
        source_counts = await conn.fetch("""
            SELECT language, COUNT(*) as count
            FROM source_texts
            GROUP BY language
            ORDER BY count DESC
        """)
        for r in source_counts:
            report.source_counts[r['language'] or 'unknown'] = int(r['count'])
        report.total_sources = sum(report.source_counts.values())

        # 4. Embedding coverage
        print('Checking embedding coverage...')
        embed_count = await conn.fetchrow("SELECT COUNT(*) as count FROM embeddings")
        report.embedding_coverage['total'] = report.total_sources
        report.embedding_coverage['with_embeddings'] = int(embed_count['count'])
        report.embedding_coverage['missing'] = report.total_sources - report.embedding_coverage['with_embeddings']
        if report.total_sources > 0:
            report.embedding_coverage['coverage_percent'] = (
                report.embedding_coverage['with_embeddings'] / report.total_sources * 100
            )

        # 5. Translation coverage
        print('Checking translation coverage...')
        trans_count = await conn.fetchrow("SELECT COUNT(*) as count FROM translations")
        report.translation_coverage['total'] = int(trans_count['count'])

        trans_with_embed = await conn.fetchrow(
            "SELECT COUNT(*) as count FROM translations WHERE embedding IS NOT NULL"
        )
        report.translation_coverage['with_embeddings'] = int(trans_with_embed['count'])

        # Check for URN-like columns
        urn_cols = await conn.fetch("""
            SELECT column_name FROM information_schema.columns
            WHERE table_name = 'translations'
            AND column_name IN ('urn', 'source_text_id', 'passage_urn', 'source_urn')
        """)
        if urn_cols:
            urn_col = urn_cols[0]['column_name']
            try:
                unique_urns = await conn.fetchrow(f"""
                    SELECT COUNT(DISTINCT {urn_col}) as count
                    FROM translations WHERE {urn_col} IS NOT NULL
                """)
                report.translation_coverage['unique_urns'] = int(unique_urns['count'])

                multi_trans = await conn.fetchrow(f"""
                    SELECT COUNT(*) as count FROM (
                        SELECT {urn_col} FROM translations
                        WHERE {urn_col} IS NOT NULL
                        GROUP BY {urn_col}
                        HAVING COUNT(*) >= 2
                    ) sub
                """)
                report.translation_coverage['multi_translation_urns'] = int(multi_trans['count'])
            except Exception as e:
                print(f'Note: URN analysis failed: {e}')

        # 6. Morphology coverage
        print('Checking morphology coverage...')
        try:
            morph_count = await conn.fetchrow("SELECT COUNT(*) as count FROM morph_entries")
            report.morphology_coverage['morph_entries'] = int(morph_count['count'])
        except:
            pass

        try:
            token_count = await conn.fetchrow("SELECT COUNT(*) as count FROM passage_tokens")
            report.morphology_coverage['passage_tokens'] = int(token_count['count'])
        except:
            pass

        # 7. Backfill jobs check
        print('Checking backfill_jobs table...')
        backfill_check = await conn.fetchrow("""
            SELECT EXISTS (
                SELECT 1 FROM information_schema.tables
                WHERE table_name = 'backfill_jobs'
            ) as exists
        """)
        report.backfill_jobs_exists = backfill_check['exists']

        # 8. Index check
        print('Checking indexes...')
        indexes = await conn.fetch("""
            SELECT indexname, indexdef
            FROM pg_indexes
            WHERE schemaname = 'public'
            AND (indexdef ILIKE '%vector%'
                 OR indexdef ILIKE '%hnsw%'
                 OR indexdef ILIKE '%ivfflat%'
                 OR indexname ILIKE '%urn%'
                 OR indexname ILIKE '%language%')
            ORDER BY indexname
        """)
        report.indexes = [{'name': r['indexname'], 'definition': r['indexdef']} for r in indexes]

        # 9. Existing translation system tables
        print('Checking existing translation system tables...')
        existing_tables = await conn.fetch("""
            SELECT table_name
            FROM information_schema.tables
            WHERE table_schema = 'public'
            AND table_name IN (
                'passage_consensus',
                'passage_style_variants',
                'translation_memory_lexeme',
                'translation_memory_phrase',
                'translation_memory_idiom',
                'translation_order_templates',
                'translation_runs',
                'translation_review_queue',
                'bridge_embeddings',
                'chunk_bridge_embeddings',
                'concept_clusters',
                'concept_members',
                'concept_edges',
                'evidence_trails'
            )
        """)
        report.existing_translation_tables = [r['table_name'] for r in existing_tables]

        # Determine primary embedding store
        if any(c['table'] == 'embeddings' and c['column'] == 'embedding'
               for c in report.embedding_infra['columns']):
            report.embedding_infra['primary_store'] = 'embeddings table'
        elif any(c['table'] == 'source_texts' and c['column'] == 'embedding'
                 for c in report.embedding_infra['columns']):
            report.embedding_infra['primary_store'] = 'source_texts.embedding'

        # 10. Check embeddings table structure
        embed_cols_detail = await conn.fetch("""
            SELECT column_name, data_type
            FROM information_schema.columns
            WHERE table_name = 'embeddings'
            ORDER BY ordinal_position
        """)
        report.embedding_infra['embeddings_table_columns'] = [
            {'column': r['column_name'], 'type': r['data_type']}
            for r in embed_cols_detail
        ]

        # 11. Sample translation structure
        trans_cols = await conn.fetch("""
            SELECT column_name, data_type
            FROM information_schema.columns
            WHERE table_name = 'translations'
            ORDER BY ordinal_position
        """)
        report.translation_coverage['table_columns'] = [
            r['column_name'] for r in trans_cols
        ]

        # Generate recommendations
        if report.embedding_coverage['coverage_percent'] < 100:
            report.recommendations.append(
                f"Embeddings coverage is {report.embedding_coverage['coverage_percent']:.1f}%. "
                f"{report.embedding_coverage['missing']:,} passages need embeddings."
            )
        if not report.backfill_jobs_exists:
            report.recommendations.append('Create backfill_jobs table for resumable processing')
        if report.translation_coverage['multi_translation_urns'] > 0:
            report.recommendations.append(
                f"{report.translation_coverage['multi_translation_urns']:,} passages have multiple translations - "
                f"eligible for consensus building"
            )
        if not any('hnsw' in idx['definition'].lower() for idx in report.indexes):
            report.recommendations.append('Consider adding HNSW indexes for vector similarity search')

    finally:
        await conn.close()

    return report


def format_report(report: AuditReport) -> str:
    """Format audit report as human-readable text."""
    output = f"""
================================================================================
                        LOGOS DATABASE AUDIT REPORT
                        {report.timestamp}
================================================================================

EMBEDDING INFRASTRUCTURE
-------------------------
Primary Store: {report.embedding_infra['primary_store']}

Vector Columns Found: {len(report.embedding_infra['columns'])}
"""

    # Group columns by table
    table_groups: Dict[str, List[str]] = {}
    for c in report.embedding_infra['columns']:
        table = c['table']
        if table not in table_groups:
            table_groups[table] = []
        table_groups[table].append(f"{c['column']} ({c['type']})")

    for table, cols in sorted(table_groups.items()):
        output += f"  {table}: {', '.join(cols)}\n"

    output += "\nDetected Dimensions:\n"
    for key, dims in report.embedding_infra['dimensions'].items():
        output += f"  {key}: {dims if dims else 'N/A'}\n"

    if 'embeddings_table_columns' in report.embedding_infra:
        output += "\nEmbeddings Table Structure:\n"
        for col in report.embedding_infra['embeddings_table_columns']:
            output += f"  {col['column']}: {col['type']}\n"

    output += f"""
SOURCE TEXT COUNTS
------------------
Total: {report.total_sources:,}
"""
    for lang, count in sorted(report.source_counts.items(), key=lambda x: -x[1]):
        output += f"  {lang}: {count:,}\n"

    output += f"""
EMBEDDING COVERAGE
------------------
Total Passages: {report.embedding_coverage['total']:,}
With Embeddings: {report.embedding_coverage['with_embeddings']:,}
Missing: {report.embedding_coverage['missing']:,}
Coverage: {report.embedding_coverage['coverage_percent']:.2f}%

TRANSLATION COVERAGE
--------------------
Total Translations: {report.translation_coverage['total']:,}
With Embeddings: {report.translation_coverage['with_embeddings']:,}
Unique URNs: {report.translation_coverage.get('unique_urns', 'N/A')}
Multi-Translation URNs (>=2): {report.translation_coverage.get('multi_translation_urns', 'N/A')}
"""

    if 'table_columns' in report.translation_coverage:
        output += f"Table Columns: {', '.join(report.translation_coverage['table_columns'])}\n"

    output += f"""
MORPHOLOGY COVERAGE
-------------------
Morph Entries: {report.morphology_coverage['morph_entries']:,}
Passage Tokens: {report.morphology_coverage['passage_tokens']:,}

INFRASTRUCTURE STATUS
---------------------
backfill_jobs table exists: {'YES' if report.backfill_jobs_exists else 'NO'}
Existing translation system tables: {', '.join(report.existing_translation_tables) if report.existing_translation_tables else 'NONE'}

VECTOR INDEXES
--------------
"""
    if not report.indexes:
        output += "  No vector/urn/language indexes found\n"
    else:
        for idx in report.indexes:
            output += f"  {idx['name']}\n"

    output += """
RECOMMENDATIONS
---------------
"""
    if not report.recommendations:
        output += "  None - infrastructure looks complete\n"
    else:
        for i, rec in enumerate(report.recommendations, 1):
            output += f"  {i}. {rec}\n"

    output += """
================================================================================
                              END OF AUDIT REPORT
================================================================================
"""
    return output


async def main():
    print('Starting LOGOS Database Audit...\n')

    try:
        report = await run_audit()
        formatted_report = format_report(report)

        # Ensure logs directory exists
        logs_dir = os.path.join(os.path.dirname(__file__), '..', '..', 'logs')
        os.makedirs(logs_dir, exist_ok=True)

        # Write text report
        audit_path = os.path.join(logs_dir, 'audit.txt')
        with open(audit_path, 'w') as f:
            f.write(formatted_report)

        # Write JSON report
        json_path = os.path.join(logs_dir, 'audit.json')
        with open(json_path, 'w') as f:
            json.dump(asdict(report), f, indent=2, default=str)

        print(formatted_report)
        print(f"\nAudit report written to: {audit_path}")
        print(f"JSON report written to: {json_path}")

    except Exception as e:
        print(f'Audit failed: {e}')
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == '__main__':
    asyncio.run(main())

File: scripts/audit/audit.ts
#!/usr/bin/env npx ts-node
/**
 * LOGOS Database Audit Script
 * Phase 0: Detect existing infrastructure and generate report
 */

import { Pool } from 'pg';
import * as fs from 'fs';
import * as path from 'path';

const DATABASE_URL = process.env.DATABASE_URL ||
  'postgresql://postgres:voqQePIH4adopQUa-1UUaFKnOT-mtsod@maglev.proxy.rlwy.net:49514/railway';

interface AuditReport {
  timestamp: string;
  embeddingInfra: {
    columns: Array<{table: string; column: string; type: string}>;
    dimensions: Record<string, number | null>;
    primaryStore: string;
  };
  sourceCounts: Record<string, number>;
  totalSources: number;
  embeddingCoverage: {
    total: number;
    withEmbeddings: number;
    missing: number;
    coveragePercent: number;
  };
  translationCoverage: {
    total: number;
    withEmbeddings: number;
    uniqueUrns: number;
    multiTranslationUrns: number;
  };
  morphologyCoverage: {
    morphEntries: number;
    passageTokens: number;
  };
  backfillJobsExists: boolean;
  indexes: Array<{name: string; definition: string}>;
  existingTranslationTables: string[];
  recommendations: string[];
}

async function runAudit(): Promise<AuditReport> {
  const pool = new Pool({
    connectionString: DATABASE_URL,
    ssl: { rejectUnauthorized: false },
    connectionTimeoutMillis: 30000,
  });

  const report: AuditReport = {
    timestamp: new Date().toISOString(),
    embeddingInfra: { columns: [], dimensions: {}, primaryStore: 'unknown' },
    sourceCounts: {},
    totalSources: 0,
    embeddingCoverage: { total: 0, withEmbeddings: 0, missing: 0, coveragePercent: 0 },
    translationCoverage: { total: 0, withEmbeddings: 0, uniqueUrns: 0, multiTranslationUrns: 0 },
    morphologyCoverage: { morphEntries: 0, passageTokens: 0 },
    backfillJobsExists: false,
    indexes: [],
    existingTranslationTables: [],
    recommendations: [],
  };

  try {
    console.log('Connecting to database...');
    const client = await pool.connect();

    // 1. Embedding columns
    console.log('Auditing embedding infrastructure...');
    const embedCols = await client.query(`
      SELECT table_name, column_name, data_type
      FROM information_schema.columns
      WHERE table_schema = 'public'
      AND (data_type ILIKE '%vector%'
           OR column_name ILIKE '%embed%'
           OR column_name ILIKE '%vector%')
      ORDER BY table_name, column_name
    `);
    report.embeddingInfra.columns = embedCols.rows.map(r => ({
      table: r.table_name,
      column: r.column_name,
      type: r.data_type
    }));

    // 2. Check vector dimensions on translations
    try {
      const transDims = await client.query(`
        SELECT vector_dims(embedding) as dims
        FROM translations
        WHERE embedding IS NOT NULL
        LIMIT 1
      `);
      if (transDims.rows.length > 0) {
        report.embeddingInfra.dimensions.translations = transDims.rows[0].dims;
      }
    } catch (e) {
      console.log('Note: translations.embedding dimension check failed');
    }

    // 3. Check style_invariant_embeddings dimensions
    try {
      const sieDims = await client.query(`
        SELECT
          CASE WHEN original_embedding IS NOT NULL THEN vector_dims(original_embedding) ELSE NULL END as orig_dims,
          CASE WHEN invariant_embedding IS NOT NULL THEN vector_dims(invariant_embedding) ELSE NULL END as inv_dims
        FROM style_invariant_embeddings
        WHERE original_embedding IS NOT NULL OR invariant_embedding IS NOT NULL
        LIMIT 1
      `);
      if (sieDims.rows.length > 0) {
        report.embeddingInfra.dimensions.style_invariant_original = sieDims.rows[0].orig_dims;
        report.embeddingInfra.dimensions.style_invariant_invariant = sieDims.rows[0].inv_dims;
      }
    } catch (e) {
      console.log('Note: style_invariant_embeddings dimension check failed');
    }

    // 4. Check passages dimensions
    try {
      const passDims = await client.query(`
        SELECT vector_dims(embedding) as dims
        FROM passages
        WHERE embedding IS NOT NULL
        LIMIT 1
      `);
      if (passDims.rows.length > 0) {
        report.embeddingInfra.dimensions.passages = passDims.rows[0].dims;
      }
    } catch (e) {
      console.log('Note: passages.embedding dimension check failed');
    }

    // 5. Source text counts by language
    console.log('Counting source texts...');
    const sourceCounts = await client.query(`
      SELECT language, COUNT(*) as count
      FROM source_texts
      GROUP BY language
      ORDER BY count DESC
    `);
    sourceCounts.rows.forEach(r => {
      report.sourceCounts[r.language || 'unknown'] = parseInt(r.count);
    });
    report.totalSources = Object.values(report.sourceCounts).reduce((a, b) => a + b, 0);

    // 6. Embedding coverage
    console.log('Checking embedding coverage...');
    const embedCount = await client.query(`SELECT COUNT(*) as count FROM embeddings`);
    report.embeddingCoverage.total = report.totalSources;
    report.embeddingCoverage.withEmbeddings = parseInt(embedCount.rows[0].count);
    report.embeddingCoverage.missing = report.totalSources - report.embeddingCoverage.withEmbeddings;
    report.embeddingCoverage.coveragePercent =
      (report.embeddingCoverage.withEmbeddings / report.totalSources * 100);

    // 7. Translation coverage
    console.log('Checking translation coverage...');
    const transCount = await client.query(`SELECT COUNT(*) as count FROM translations`);
    report.translationCoverage.total = parseInt(transCount.rows[0].count);

    const transWithEmbed = await client.query(`
      SELECT COUNT(*) as count FROM translations WHERE embedding IS NOT NULL
    `);
    report.translationCoverage.withEmbeddings = parseInt(transWithEmbed.rows[0].count);

    // Check for URN column and count unique
    try {
      const urnCheck = await client.query(`
        SELECT column_name FROM information_schema.columns
        WHERE table_name = 'translations' AND column_name IN ('urn', 'source_text_id', 'passage_urn')
      `);
      if (urnCheck.rows.length > 0) {
        const urnCol = urnCheck.rows[0].column_name;
        const uniqueUrns = await client.query(`
          SELECT COUNT(DISTINCT ${urnCol}) as count FROM translations WHERE ${urnCol} IS NOT NULL
        `);
        report.translationCoverage.uniqueUrns = parseInt(uniqueUrns.rows[0].count);

        const multiTrans = await client.query(`
          SELECT COUNT(*) as count FROM (
            SELECT ${urnCol} FROM translations
            WHERE ${urnCol} IS NOT NULL
            GROUP BY ${urnCol}
            HAVING COUNT(*) >= 2
          ) sub
        `);
        report.translationCoverage.multiTranslationUrns = parseInt(multiTrans.rows[0].count);
      }
    } catch (e) {
      console.log('Note: URN analysis failed');
    }

    // 8. Morphology coverage
    console.log('Checking morphology coverage...');
    try {
      const morphCount = await client.query(`SELECT COUNT(*) as count FROM morph_entries`);
      report.morphologyCoverage.morphEntries = parseInt(morphCount.rows[0].count);
    } catch (e) {
      report.morphologyCoverage.morphEntries = 0;
    }

    try {
      const tokenCount = await client.query(`SELECT COUNT(*) as count FROM passage_tokens`);
      report.morphologyCoverage.passageTokens = parseInt(tokenCount.rows[0].count);
    } catch (e) {
      report.morphologyCoverage.passageTokens = 0;
    }

    // 9. Backfill jobs check
    console.log('Checking backfill_jobs table...');
    const backfillCheck = await client.query(`
      SELECT EXISTS (
        SELECT 1 FROM information_schema.tables
        WHERE table_name = 'backfill_jobs'
      ) as exists
    `);
    report.backfillJobsExists = backfillCheck.rows[0].exists;

    // 10. Index check
    console.log('Checking indexes...');
    const indexes = await client.query(`
      SELECT indexname, indexdef
      FROM pg_indexes
      WHERE schemaname = 'public'
      AND (indexdef ILIKE '%vector%'
           OR indexdef ILIKE '%hnsw%'
           OR indexdef ILIKE '%ivfflat%'
           OR indexname ILIKE '%urn%'
           OR indexname ILIKE '%language%')
      ORDER BY indexname
    `);
    report.indexes = indexes.rows.map(r => ({
      name: r.indexname,
      definition: r.indexdef
    }));

    // 11. Existing translation system tables
    console.log('Checking existing translation system tables...');
    const existingTables = await client.query(`
      SELECT table_name
      FROM information_schema.tables
      WHERE table_schema = 'public'
      AND table_name IN (
        'passage_consensus',
        'passage_style_variants',
        'translation_memory_lexeme',
        'translation_memory_phrase',
        'translation_memory_idiom',
        'translation_order_templates',
        'translation_runs',
        'translation_review_queue',
        'bridge_embeddings',
        'chunk_bridge_embeddings',
        'concept_clusters',
        'concept_members',
        'concept_edges',
        'evidence_trails'
      )
    `);
    report.existingTranslationTables = existingTables.rows.map(r => r.table_name);

    // Determine primary embedding store
    if (report.embeddingInfra.columns.some(c => c.table === 'embeddings' && c.column === 'embedding')) {
      report.embeddingInfra.primaryStore = 'embeddings table';
    } else if (report.embeddingInfra.columns.some(c => c.table === 'source_texts' && c.column === 'embedding')) {
      report.embeddingInfra.primaryStore = 'source_texts.embedding';
    }

    // Generate recommendations
    if (report.embeddingCoverage.coveragePercent < 100) {
      report.recommendations.push(
        `Embeddings coverage is ${report.embeddingCoverage.coveragePercent.toFixed(1)}%. ` +
        `${report.embeddingCoverage.missing.toLocaleString()} passages need embeddings.`
      );
    }
    if (!report.backfillJobsExists) {
      report.recommendations.push('Create backfill_jobs table for resumable processing');
    }
    if (report.translationCoverage.multiTranslationUrns > 0) {
      report.recommendations.push(
        `${report.translationCoverage.multiTranslationUrns.toLocaleString()} passages have multiple translations - ` +
        `eligible for consensus building`
      );
    }
    if (report.indexes.filter(i => i.definition.includes('hnsw')).length === 0) {
      report.recommendations.push('Consider adding HNSW indexes for vector similarity search');
    }

    client.release();
    return report;

  } finally {
    await pool.end();
  }
}

function formatReport(report: AuditReport): string {
  let output = `
================================================================================
                        LOGOS DATABASE AUDIT REPORT
                        ${report.timestamp}
================================================================================

EMBEDDING INFRASTRUCTURE
-------------------------
Primary Store: ${report.embeddingInfra.primaryStore}

Vector Columns Found: ${report.embeddingInfra.columns.length}
`;

  const tableGroups: Record<string, string[]> = {};
  report.embeddingInfra.columns.forEach(c => {
    if (!tableGroups[c.table]) tableGroups[c.table] = [];
    tableGroups[c.table].push(`${c.column} (${c.type})`);
  });
  Object.entries(tableGroups).forEach(([table, cols]) => {
    output += `  ${table}: ${cols.join(', ')}\n`;
  });

  output += `
Detected Dimensions:
`;
  Object.entries(report.embeddingInfra.dimensions).forEach(([key, dims]) => {
    output += `  ${key}: ${dims ?? 'N/A'}\n`;
  });

  output += `
SOURCE TEXT COUNTS
------------------
Total: ${report.totalSources.toLocaleString()}
`;
  Object.entries(report.sourceCounts)
    .sort((a, b) => b[1] - a[1])
    .forEach(([lang, count]) => {
      output += `  ${lang}: ${count.toLocaleString()}\n`;
    });

  output += `
EMBEDDING COVERAGE
------------------
Total Passages: ${report.embeddingCoverage.total.toLocaleString()}
With Embeddings: ${report.embeddingCoverage.withEmbeddings.toLocaleString()}
Missing: ${report.embeddingCoverage.missing.toLocaleString()}
Coverage: ${report.embeddingCoverage.coveragePercent.toFixed(2)}%

TRANSLATION COVERAGE
--------------------
Total Translations: ${report.translationCoverage.total.toLocaleString()}
With Embeddings: ${report.translationCoverage.withEmbeddings.toLocaleString()}
Unique URNs: ${report.translationCoverage.uniqueUrns.toLocaleString()}
Multi-Translation URNs (>=2): ${report.translationCoverage.multiTranslationUrns.toLocaleString()}

MORPHOLOGY COVERAGE
-------------------
Morph Entries: ${report.morphologyCoverage.morphEntries.toLocaleString()}
Passage Tokens: ${report.morphologyCoverage.passageTokens.toLocaleString()}

INFRASTRUCTURE STATUS
---------------------
backfill_jobs table exists: ${report.backfillJobsExists ? 'YES' : 'NO'}
Existing translation system tables: ${report.existingTranslationTables.length > 0 ? report.existingTranslationTables.join(', ') : 'NONE'}

VECTOR INDEXES
--------------
`;
  if (report.indexes.length === 0) {
    output += '  No vector/urn/language indexes found\n';
  } else {
    report.indexes.forEach(idx => {
      output += `  ${idx.name}\n`;
    });
  }

  output += `
RECOMMENDATIONS
---------------
`;
  if (report.recommendations.length === 0) {
    output += '  None - infrastructure looks complete\n';
  } else {
    report.recommendations.forEach((rec, i) => {
      output += `  ${i + 1}. ${rec}\n`;
    });
  }

  output += `
================================================================================
                              END OF AUDIT REPORT
================================================================================
`;
  return output;
}

async function main() {
  console.log('Starting LOGOS Database Audit...\n');

  try {
    const report = await runAudit();
    const formattedReport = formatReport(report);

    // Write to logs/audit.txt
    const logsDir = path.join(__dirname, '..', '..', 'logs');
    if (!fs.existsSync(logsDir)) {
      fs.mkdirSync(logsDir, { recursive: true });
    }
    const auditPath = path.join(logsDir, 'audit.txt');
    fs.writeFileSync(auditPath, formattedReport);

    // Also write JSON for programmatic access
    const jsonPath = path.join(logsDir, 'audit.json');
    fs.writeFileSync(jsonPath, JSON.stringify(report, null, 2));

    console.log(formattedReport);
    console.log(`\nAudit report written to: ${auditPath}`);
    console.log(`JSON report written to: ${jsonPath}`);

  } catch (error) {
    console.error('Audit failed:', error);
    process.exit(1);
  }
}

main();
